### PR TITLE
feat: add `preChecks`

### PR DESCRIPTION
**Pull request by 0z13**
The idea is to have healthchecks in the pre-activation stage aswell, this is useful if you wanna make sure there are no active connections to the node you are about to switch. 

the api/logic is exactly like current healthchecks, except no `deployment.healthChecks.http` style checks.

example: 
```
  deployment.preChecks.cmd =
    [
      {
        cmd = [
          "false"
        ];
        description = "Node is quarantined, do not touch.";
      }
    ];
```

### CODE REVIEW

File: data/options.nix [hunk 1]
```diff
@@ -92,6 +92,19 @@ let
     };
   });
 
+  preCheckType = submodule (_: {
```
Comment (adamtulinius) [inline data/options.nix:95]: I think it's should be possible to reuse the existing `healthCheckType` here

File: healthchecks/types.go [hunk 2]
```diff
@@ -137,3 +150,34 @@ func (healthCheck HttpHealthCheck) Run(host Host) error {
...
+
+func (pre PreActivationCheck) GetDescription() string {
+	return pre.Description
+}
+
+func (pre PreActivationCheck) GetPeriod() int {
+	return pre.Period
+}
+
+func (pre PreActivationCheck) Run(host Host) error {
```
Comment (adamtulinius) [inline healthchecks/types.go:162]: Ideally we could merge this with `Run` from the old health checks, while still using these new error messages

File: healthchecks/healthchecks.go [hunk 3]
```diff
@@ -68,3 +68,42 @@ func runCheckUntilSuccess(host Host, healthCheck HealthCheck, wg *sync.WaitGroup
 	}
 	wg.Done()
 }
+
+func PerformPreChecks(sshContext *ssh.SSHContext, host Host, timeout int) (err error) {
```
Comment (adamtulinius) [inline healthchecks/healthchecks.go:72]: Let's try to merge this with `Perform`, and along with the type change we should get http checks as pre-checks for free

**Review by adamtulinius — CHANGES_REQUESTED** [review]
tl;dr: I would like to try merging the code with the existing health checks, and make all check types supported as pre checks as well

**Comment by adamtulinius** [comment]
I wonder if `preChecks` should be named so it's more clear that it's pre-deploy checks. On the other hand it's also implicit that health checks are post-deploy. Thoughts? When we pick a name it's hard to change down the line. I'm probably just rambling here.

File: healthchecks/types.go (hunk 2 above)
Comment (adamtulinius) [reply healthchecks/types.go:162]: Maybe it's the calling function (`runCheckUntilSuccess`?) could prefix the health check errors with e.g. `Pre-activation check error`, or maybe it'll be too much of a hassle

**Comment by cafkafk** [comment]
> I wonder if `preChecks` should be named so it's more clear that it's pre-deploy checks. On the other hand it's also implicit that health checks are post-deploy. Thoughts? When we pick a name it's hard to change down the line. I'm probably just rambling here.

pre-deploy-checks (or, well `preDeployChecks`) are better IMO :+1: 

<del>I'd also support renaming healthchecks</del>

**Comment by adamtulinius** [comment]
> > I wonder if `preChecks` should be named so it's more clear that it's pre-deploy checks. On the other hand it's also implicit that health checks are post-deploy. Thoughts? When we pick a name it's hard to change down the line. I'm probably just rambling here.
> 
> pre-deploy-checks (or, well `preDeployChecks`) are better IMO 👍
> 
> ~I'd also support renaming healthchecks~

I wonder if the semantics are actually different enough to call it something completely else. Some CI tools use the concept of deployment gates, and I wonder if we should use that here as well. That way we can also open up the question if morph should wait for a gate to become true/open, or if we would want morph to exit when the requirements for a gate is not met.

**Comment by adamtulinius** [comment]
What I have done :scream::
1. rebased on master
2. renamed preCheck to preDeployCheck
3. removed all the code duplication and extra nix types

This have caused the following side-effects (maybe I missed something?):
1. preDeployChecks now support both commands and HTTP requests (exactly like health checks)
2. The output of errors from running health checks have had some superfluous text cut from error messages
3. Morph will now print "Running health checks on [...]" instead of "Running healthchecks on [...]"


The only problem I see is if someone is parsing the human readable output from morph, but we never promised that to be stable, so I think it's worth changing the output slightly to avoid copy/pasting code.

I'm not sure that preDeployChecks should keep running forever (technically right now they obey the `--timeout` flag, which is shared with healthChecks). Maybe we want them to actually just cause morph to fail instantly. Thoughts?

@0z13 - can you kindly test that this is still working, and review my changes? Since we've both touched this MR I suspect we will need a third person to approve, but I want to hear your thoughts.

**Comment by 0z13** [comment]
beautiful diff stats on that commit @adamtulinius, this is soo much nicer now :D

**Comment by srhb** [comment]
I understand you need a tie-breaker for the merge. Has this been tested anywhere?

**Comment by adamtulinius** [comment]
> I understand you need a tie-breaker for the merge. Has this been tested anywhere?

I hacked an existing deployment (we really need automated tests, heh) to test the new functionality, and tested that normal health checks still work (albeit with slightly altered output). @0z13 also did testing, but only wrote about it on Slack (I'll DM you).

**Comment by srhb** [comment]
Oh, ok, works for me. And yes to automated tests, I'll write a story.

**Review by srhb — APPROVED** [review]
Dummy approval, participants did the actual cross-review
